### PR TITLE
Python: Modernise and fix FP in `py/undefined-export`

### LINKED ISSUE #1899: Python FP in py/undefined-export
[Occurs when imported names are exported](https://lgtm.com/projects/g/dials/dials/snapshot/5c386da2f97d98b7ec80e34cec0cbf894fce6a36/files/algorithms/refinement/parameterisation/__init__.py?sort=name&dir=ASC&mode=heatmap):

image

Note that only 2 of the 3 names are marked as undefined. The name from the second import statement was not flagged.

### LINKED ISSUE #2071: LGTM.com - false positive "Explicit export is not defined"
**Description of the false positive**
I think there is a false positive on "Explicit export is not defined"
I have already defined the function but it tells that "the name exported by __all__ is not defined". I tested that the function can be imported outside using import * so I think there should be no problem.

**URL to the alert on the project page on LGTM.com**
https://lgtm.com/projects/g/apache/incubator-singa/snapshot/e42fca51ff4c03b456ceff12974dba64934a86be/files/examples/autograd/xceptionnet.py?sort=name&dir=ASC&mode=heatmap#x7cb181f811af62ad:1

### PR DESCRIPTION
**Pull request by taus-semmle**
This should fix #2071. (Edit: and fix #1899.)

A few API changes were necessary in order to modernise the query. I decided to add `getShortName` (already present on `ModuleObject`) to `Module` and not `ModuleValue`, as this seemed to be the better place for it. (I could only find a single reference to `ModuleObject::getShortName` -- the one in this query.)

The second change is the addition of `Value::isAbsent`. This one I'm a bit more conflicted about. Previously, lack of points-to information was a good indicator that our analysis was incomplete, and that this might lead to false positives. This no longer works, as we can have a `ModuleValue` for a module that doesn't exist in source code, but which probably _ought_ to exist because of how it is used in the source code. I welcome any comments about the name of the method and its QLDoc string. 

As this FP was essentially fixed by modernising the query, I do not think it merits a change note. Rather than adding a change note for each such modernised query (there will be many of these), I think it will be more useful to just add an overall section to the change notes describing the modernisation effort.

### CODE REVIEW

**Comment by taus-semmle** [comment]
Also fixes #1899.

**Comment by taus-semmle** [comment]
Changed the second commit to also account for the case where the points-to information is missing altogether, and not just "absent".

File: python/ql/src/Variables/UndefinedExport.ql [hunk 1]
```diff
@@ -14,39 +14,53 @@
...
+        a.getScope() = m and
+        all.getId() = "__all__" and
+        a.getValue().(List).getAnElt() = name
     )
 }
 
-predicate mutates_globals(PythonModuleObject m) {
+predicate mutates_globals(ModuleValue m) {
     exists(CallNode globals |
         globals = Object::builtin("globals").(FunctionObject).getACall() and
```
Comment (RasmusWL) [inline python/ql/src/Variables/UndefinedExport.ql]: `FunctionObject` => `FunctionValue`?

File: python/ql/src/Variables/UndefinedExport.ql [hunk 2]
```diff
@@ -14,39 +14,53 @@
...
         exists(SubscriptNode sub | sub.getValue() = globals and sub.isStore())
     )
     or
     exists(Object enum_convert |
         enum_convert.hasLongName("enum.Enum._convert") and
-        exists(CallNode call |
-            call.getScope() = m.getModule()
-            |
+        exists(CallNode call | call.getScope() = m.getScope() |
             enum_convert.(FunctionObject).getACall() = call or
```
Comment (RasmusWL) [inline python/ql/src/Variables/UndefinedExport.ql]: `FunctionObject` => `FunctionValue`?

File: python/ql/src/Variables/UndefinedExport.ql [hunk 3]
```diff
@@ -14,39 +14,53 @@
...
             call.getFunction().refersTo(enum_convert)
```
Comment (RasmusWL) [inline python/ql/src/Variables/UndefinedExport.ql]: refersTo => pointsTo

File: python/ql/src/Variables/UndefinedExport.ql (hunk 2 above)
Comment (tausbn) [reply python/ql/src/Variables/UndefinedExport.ql]: Done.

File: python/ql/src/Variables/UndefinedExport.ql (hunk 3 above)
Comment (tausbn) [reply python/ql/src/Variables/UndefinedExport.ql]: Done.

File: python/ql/src/Variables/UndefinedExport.ql (hunk 1 above)
Comment (tausbn) [reply python/ql/src/Variables/UndefinedExport.ql]: Done.

**Comment by tausbn** [comment]
I fixed the missing modernisation. My attempts at creating a small example of the false positive failed (as in, they were not flagged by the query), so I'm not entirely sure what's going on there. I did verify that the mentioned false positives are fixed (by running the new query on downloaded snapshots), however.

**Pull request merged**